[threaded-animations] scroll-driven animations don't update in sync with scroll updates on iOS

graouts opened the pull request:
#### 230a3fd2ec731e03af1199f9fdb712c603de333f
<pre>
[threaded-animations] scroll-driven animations don&apos;t update in sync with scroll updates on iOS
<a href="https://bugs.webkit.org/show_bug.cgi?id=304189">https://bugs.webkit.org/show_bug.cgi?id=304189</a>
<a href="https://rdar.apple.com/166546662">rdar://166546662</a>

Reviewed by Anne van Kesteren.

On iOS, we would update all threaded animations, scroll-driven as well as time-based, as the
display refreshes. However, scroll-driven animations should be updated as the remote scrolling
three updates, and time-based animations should be updated as the display is updated.

We already update progress-based timelines correctly as the remote scrolling tree updates,
in `RemoteScrollingTreeIOS::scrollingTreeNodeDidScroll()`. That method already call
`RemoteScrollingCoordinatorProxyIOS::progressBasedTimelinesWereUpdatedForNode()`, so now
instead of scheduling a display refresh callback we call the new method
`updateAnimationStacksDependentOnScrollingNode()` such that all animations stacks that
have animations associated with a timeline that as that scrolling node as its source
get updated.

We also introduce a new `updateTimeDependentAnimationStacks()` method which only updates
animation stacks that contain animations associated with a monotonic timeline, and only
call this method under `displayDidRefresh()`.

Finally, we must make sure to only schedule display refresh callbacks for monotonic animations,
so we rename the various methods on `RemoteLayerTreeDrawingAreaProxyIOS` that are related
scheduling such updates to make it clear they&apos;re only relevant to monotonic animations, and then
we only call them in situations where we have animations stacks containing monotonic animations.

As a result, we update progress-based timelines and any animation stack containing an associated
scroll-driven animation as a remote scrolling tree node is updated, and we schedule display refresh
callbacks for animations only if we have monotonic animations, only updating animation stacks
containing such animations as those callbacks fire.

As a result we can remove `RemoteScrollingTree::hasTimelineForNode()` and the methods under it
since it is now unused.

* Source/WebKit/UIProcess/RemoteLayerTree/RemoteAnimationStack.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteAnimationStack.mm:
(WebKit::RemoteAnimationStack::isDependentOnScrollingNodeWithID const):
(WebKit::RemoteAnimationStack::isTimeDependent const):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteAnimationTimeline.h:
(WebKit::RemoteAnimationTimeline::isMonotonic const):
(WebKit::RemoteAnimationTimeline::isProgressBased const):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteMonotonicTimeline.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteProgressBasedTimeline.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteProgressBasedTimelineRegistry.cpp:
(WebKit::RemoteProgressBasedTimelineRegistry::hasTimelineForNode const): Deleted.
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteProgressBasedTimelineRegistry.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingTree.cpp:
(WebKit::RemoteScrollingTree::hasTimelineForNode const): Deleted.
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingTree.h:
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteLayerTreeDrawingAreaProxyIOS.h:
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteLayerTreeDrawingAreaProxyIOS.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxyIOS::didRefreshDisplay):
(WebKit::RemoteLayerTreeDrawingAreaProxyIOS::pauseDisplayRefreshCallbacks):
(WebKit::RemoteLayerTreeDrawingAreaProxyIOS::scheduleDisplayRefreshCallbacksForMonotonicAnimations):
(WebKit::RemoteLayerTreeDrawingAreaProxyIOS::pauseDisplayRefreshCallbacksForMonotonicAnimations):
(WebKit::RemoteLayerTreeDrawingAreaProxyIOS::scheduleDisplayRefreshCallbacksForAnimation): Deleted.
(WebKit::RemoteLayerTreeDrawingAreaProxyIOS::pauseDisplayRefreshCallbacksForAnimation): Deleted.
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingCoordinatorProxyIOS.h:
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingCoordinatorProxyIOS.mm:
(WebKit::RemoteScrollingCoordinatorProxyIOS::displayDidRefresh):
(WebKit::RemoteScrollingCoordinatorProxyIOS::animationsWereAddedToNode):
(WebKit::RemoteScrollingCoordinatorProxyIOS::progressBasedTimelinesWereUpdatedForNode):
(WebKit::RemoteScrollingCoordinatorProxyIOS::animationsWereRemovedFromNode):
(WebKit::RemoteScrollingCoordinatorProxyIOS::updateTimeDependentAnimationStacks):
(WebKit::RemoteScrollingCoordinatorProxyIOS::updateAnimationStacksDependentOnScrollingNode):
(WebKit::RemoteScrollingCoordinatorProxyIOS::updateAnimationStacks):
(WebKit::RemoteScrollingCoordinatorProxyIOS::updateAnimations): Deleted.

Canonical link: <a href="https://commits.webkit.org/304619@main">https://commits.webkit.org/304619@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d995da3c27e23f6e4361a8a5bf9840a42de1d0aa

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/136093 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/8449 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/47372 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/143799 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/89061 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/50566d87-e4d0-4ce8-81c9-dd78334d2659) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/9122 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/8294 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/104040 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/89061 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/8ca9e2ff-f046-4c2f-98c6-065edd421e40) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/139039 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/6632 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/121963 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/84887 "Found 1 new API test failure: WPE/TestWebKitWebContext:/webkit/WebKitWebContext/uri-scheme (failure)") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/f959bdb4-5508-4fe9-96bc-e022a12da2ff) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/6308 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/3952 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/4395 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/115585 "Passed tests") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/146547 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/8133 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/40724 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/112400 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/8149 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/6833 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/112749 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/6215 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/118263 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/62120 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20969 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/8181 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/36326 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/7896 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/71740 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/8120 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/7973 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingCoordinatorProxyIOS.mm:
```diff
@@ -708,21 +707,40 @@ static bool overlayDrawsAboveScrollView(const HashSet<RetainPtr<UIView>>& overla
+    m_monotonicTimelineRegistry->advanceCurrentTime(MonotonicTime::now());
+
+    updateAnimationStacks([](auto& animationStack) {
+        return Ref { animationStack }->isTimeDependent();
```
Why do we need to Ref it here? Isn't the caller handling that for us?

graouts replied:
You are correct, I'll remove the extra ref-churn.

annevk commented on Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingCoordinatorProxyIOS.mm:
```diff
@@ -708,21 +707,40 @@ static bool overlayDrawsAboveScrollView(const HashSet<RetainPtr<UIView>>& overla
+{
+    auto scrollingNodeID = node.scrollingNodeID();
+    updateAnimationStacks([scrollingNodeID](auto& animationStack) {
+        return Ref { animationStack }->isDependentOnScrollingNodeWithID(scrollingNodeID);
```
Ditto.

graouts committed:
[threaded-animations] scroll-driven animations don't update in sync with scroll updates on iOS
https://bugs.webkit.org/show_bug.cgi?id=304189
rdar://166546662

Reviewed by Anne van Kesteren.

On iOS, we would update all threaded animations, scroll-driven as well as time-based, as the
display refreshes. However, scroll-driven animations should be updated as the remote scrolling
three updates, and time-based animations should be updated as the display is updated.

We already update progress-based timelines correctly as the remote scrolling tree updates,
in `RemoteScrollingTreeIOS::scrollingTreeNodeDidScroll()`. That method already call
`RemoteScrollingCoordinatorProxyIOS::progressBasedTimelinesWereUpdatedForNode()`, so now
instead of scheduling a display refresh callback we call the new method
`updateAnimationStacksDependentOnScrollingNode()` such that all animations stacks that
have animations associated with a timeline that as that scrolling node as its source
get updated.

We also introduce a new `updateTimeDependentAnimationStacks()` method which only updates
animation stacks that contain animations associated with a monotonic timeline, and only
call this method under `displayDidRefresh()`.

Finally, we must make sure to only schedule display refresh callbacks for monotonic animations,
so we rename the various methods on `RemoteLayerTreeDrawingAreaProxyIOS` that are related
scheduling such updates to make it clear they're only relevant to monotonic animations, and then
we only call them in situations where we have animations stacks containing monotonic animations.

As a result, we update progress-based timelines and any animation stack containing an associated
scroll-driven animation as a remote scrolling tree node is updated, and we schedule display refresh
callbacks for animations only if we have monotonic animations, only updating animation stacks
containing such animations as those callbacks fire.

As a result we can remove `RemoteScrollingTree::hasTimelineForNode()` and the methods under it
since it is now unused.

* Source/WebKit/UIProcess/RemoteLayerTree/RemoteAnimationStack.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteAnimationStack.mm:
(WebKit::RemoteAnimationStack::isDependentOnScrollingNodeWithID const):
(WebKit::RemoteAnimationStack::isTimeDependent const):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteAnimationTimeline.h:
(WebKit::RemoteAnimationTimeline::isMonotonic const):
(WebKit::RemoteAnimationTimeline::isProgressBased const):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteMonotonicTimeline.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteProgressBasedTimeline.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteProgressBasedTimelineRegistry.cpp:
(WebKit::RemoteProgressBasedTimelineRegistry::hasTimelineForNode const): Deleted.
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteProgressBasedTimelineRegistry.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingTree.cpp:
(WebKit::RemoteScrollingTree::hasTimelineForNode const): Deleted.
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingTree.h:
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteLayerTreeDrawingAreaProxyIOS.h:
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteLayerTreeDrawingAreaProxyIOS.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxyIOS::didRefreshDisplay):
(WebKit::RemoteLayerTreeDrawingAreaProxyIOS::pauseDisplayRefreshCallbacks):
(WebKit::RemoteLayerTreeDrawingAreaProxyIOS::scheduleDisplayRefreshCallbacksForMonotonicAnimations):
(WebKit::RemoteLayerTreeDrawingAreaProxyIOS::pauseDisplayRefreshCallbacksForMonotonicAnimations):
(WebKit::RemoteLayerTreeDrawingAreaProxyIOS::scheduleDisplayRefreshCallbacksForAnimation): Deleted.
(WebKit::RemoteLayerTreeDrawingAreaProxyIOS::pauseDisplayRefreshCallbacksForAnimation): Deleted.
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingCoordinatorProxyIOS.h:
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingCoordinatorProxyIOS.mm:
(WebKit::RemoteScrollingCoordinatorProxyIOS::displayDidRefresh):
(WebKit::RemoteScrollingCoordinatorProxyIOS::animationsWereAddedToNode):
(WebKit::RemoteScrollingCoordinatorProxyIOS::progressBasedTimelinesWereUpdatedForNode):
(WebKit::RemoteScrollingCoordinatorProxyIOS::animationsWereRemovedFromNode):
(WebKit::RemoteScrollingCoordinatorProxyIOS::updateTimeDependentAnimationStacks):
(WebKit::RemoteScrollingCoordinatorProxyIOS::updateAnimationStacksDependentOnScrollingNode):
(WebKit::RemoteScrollingCoordinatorProxyIOS::updateAnimationStacks):
(WebKit::RemoteScrollingCoordinatorProxyIOS::updateAnimations): Deleted.

Canonical link: https://commits.webkit.org/304619@main
